### ISSUE TITLE
Installing via poetry

### ISSUE THREAD

**Issue by dmatos2012** [pmeier/light-the-torch#32]
Hello,
I would like to use this to install torch with poetry, but when running it as such `poetry run ltt --help` or any other command `poetry run ltt install torch`, I get

```
 File "/home/david/.cache/pypoetry/virtualenvs/torch-coco-detection-l0t67CWY-py3.6/lib/python3.6/site-packages/light_the_torch/cli/__init__.py", line 173, in add_pip_install_arguments
    assert len(options) == 1
AssertionError
```
Is this even possible? Or do I have to do get this repo as kind of poetry source? 

Btw, it works perfectly if I use it how you explain in the documentation.

Thanks by the way for this.

**Comment by pmeier** [comment]
Hey @dmatos2012 

> Is this even possible? Or do I have to do get this repo as kind of poetry source?

I don't know. I've never worked with `poetry` before. Could you post the `pyproject.toml` file that you are using? Without it, I'm not even getting to the error above.

The error itself is also weird. I was fairly certain that this would never be triggered, hence no error message. What `pip` version are you using?

**Comment by dmatos2012** [comment]
Hi @pmeier thanks for the quick reply. 
```
[tool.poetry]
name = "torch-coco-detection"
version = "0.1.0"
description = "blah"
authors = ["blah"]
readme = "README.md"
packages = [{include = "torch_coco_detection"}]

[tool.poetry.dependencies]
python = "3.6.9"
torch = "1.7.1+cu101"
torchvision = "0.8.2+cu101"
light-the-torch = "^0.3.1"

[tool.poetry.dev-dependencies]
pytest = "^6.2.4"
pytest-cov = "^2.12.1"
black = "^21.6-beta.0"
isort = "^5.9.1"
flake8 = "^3.9.2"
bandit = "^1.7.0"
safety = "^1.10.3"

[[tool.poetry.source]]
name = "torch_rep"
url = "https://eternalphane.github.io/pytorch-pypi"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```
```
pip --version
# 20.1`
```
Basically what I am trying to do is get torch, torchvision with poetry, but this [issue](https://github.com/python-poetry/poetry/issues/4124) came up.  I solved it by adding that poetry.source repo and was able to do `poetry add torch=1.7.1+cu101`. However, I was trying to get torch instead by using your tool.

So by running `poetry run ltt --help` in my project root folder, i get the whole error
Traceback (most recent call last):
```
  File "/home/david/.cache/pypoetry/virtualenvs/torch-coco-detection-l0t67CWY-py3.6/bin/ltt", line 8, in <module>
    sys.exit(main())
  File "/home/david/.cache/pypoetry/virtualenvs/torch-coco-detection-l0t67CWY-py3.6/lib/python3.6/site-packages/light_the_torch/cli/__init__.py", line 12, in main
    args = parse_args(args)
  File "/home/david/.cache/pypoetry/virtualenvs/torch-coco-detection-l0t67CWY-py3.6/lib/python3.6/site-packages/light_the_torch/cli/__init__.py", line 26, in parse_args
    parser = make_ltt_parser()
  File "/home/david/.cache/pypoetry/virtualenvs/torch-coco-detection-l0t67CWY-py3.6/lib/python3.6/site-packages/light_the_torch/cli/__init__.py", line 87, in make_ltt_parser
    add_ltt_find_parser(subparsers)
  File "/home/david/.cache/pypoetry/virtualenvs/torch-coco-detection-l0t67CWY-py3.6/lib/python3.6/site-packages/light_the_torch/cli/__init__.py", line 164, in add_ltt_find_parser
    add_pip_install_arguments(parser, "platform", "python_version")
  File "/home/david/.cache/pypoetry/virtualenvs/torch-coco-detection-l0t67CWY-py3.6/lib/python3.6/site-packages/light_the_torch/cli/__init__.py", line 173, in add_pip_install_arguments
    assert len(options) == 1
AssertionError
```
It can also be that I am doing something wrong, which in that case it would be unrelated to you but thought i would ask anyways.

**Comment by pmeier** [comment]
I still can't reproduce your error. With the `pyproject.toml` you posted above it seems to be working as expected:

```
$poetry run ltt install torch
Collecting torch==1.9.0+cu111
  Downloading https://download.pytorch.org/whl/cu111/torch-1.9.0%2Bcu111-cp39-cp39-linux_x86_64.whl (2041.4 MB)
     |▍                               | 23.1 MB 39.5 MB/s eta 0:00:52
```

Could you "start fresh" and check if this is also working for you? If not, what extra steps do I need to perform to hit your error?

**Comment by dmatos2012** [comment]
How are you installing light-the-torch? Because oddly enough, if I create  a venv, and `pip install light-the-torch` and go back to my project and do  `poetry run ltt install torch` within the venv it works. 

However, if i just make a random directory, and do `poetry add light-the-torch` to add your package, and then do `poetry run ltt` I get the previously mentioned error. 

Doing system-wide install yielded the same error. This is so strange. So sorry if I am wasting your time in this.

**Comment by pmeier** [comment]
> How are you installing light-the-torch? Because oddly enough, if I create a venv, and pip install light-the-torch and go back to my project and do poetry run ltt install torch within the venv it works.

That is also how I did it.

---

> However, if i just make a random directory, and do `poetry add light-the-torch` to add your package, and then do `poetry run ltt` I get the previously mentioned error.

If I do the same, I get

```
RuntimeError

  Poetry could not find a pyproject.toml file in /tmp/bar or its parents
```

If I copy the `pyproject.toml` from before, I get

```
The following packages are already present in the pyproject.toml and will be skipped:

  • light-the-torch

If you want to update it to the latest compatible version, you can use `poetry update package`.
If you prefer to upgrade it to the latest available version, you can use `poetry add package@latest`.

Nothing to add.
```

So IMO there are some steps missing from your description.

---

> So sorry if I am wasting your time in this.

Well, I'm happy to help you debug your issue, because it might be a problem on my end. But as stated above, I've never used `poetry` before, and I'm not up to learn all its intricacies just for debugging this. Thus, I need a full [minimal working example (MWE)](https://en.wikipedia.org/wiki/Minimal_working_example) from you that I can follow step by step to get to your bug. A good MWE takes some skill and time to compile, but believe me maintainers will really appreciate having one in a bug report.

**Comment by dmatos2012** [comment]
> RuntimeError
 Poetry could not find a pyproject.toml file in /tmp/bar or its parents

Oh sorry, I forgot to mention in my comment you had to do `poetry init` in the directory created, which creates a pyproject.toml with only python as dependency, thats why that error. It also said light-the-torch was installed because when you copied my pyproject.toml , it was already in the dependencies. 

> A good MWE takes some skill and time to compile, but believe me maintainers will really appreciate having one in a bug report.

Yeah, I dont know how to do one, as I have never provided one before. I am not that experienced with dependencies, and those kind of things. 

> But as stated above, I've never used poetry before, and I'm not up to learn all its intricacies just for debugging this.

I understand 100%, so feel free to close this issue, and perhaps if I learn how to do a proper MWE, I can reference this issue and open a new one.

Thanks for the quick replies, and willingness to help :)

**Comment by pmeier** [comment]
> Yeah, I dont know how to do one, as I have never provided one before. I am not that experienced with dependencies, and those kind of things.

In that case, lets use this as a learning opportunity. An MWE for the bug you are seeing might look like this:

1. `pip install poetry`
2. `poetry init -n`
3. `poetry add light-the-torch`
4. `ltt --help`

Lets go through it step by step:

1. `pip install poetry`: Since `poetry` is not part of this project, it is good practice to include install instructions.
2. `poetry init -n`: The `-n` flag actually has two advantages here:
    1. I don't need to go through the configuration manually.
    2. Since we only using default options, we can be fairly certain that it is not a weird configuration that is causing the problem.
3. `poetry add light-the-torch`: Nothing of particular interest here.
4. `ltt --help`: The failure is happening even without running it from `poetry`! That means just by installing it from another tool `light-the-torch` is not working properly.

With these 4 steps I now have a clear way of debugging this without ever needing to step into `poetry` land. I'll get right on it.

**Comment by pmeier** [comment]
Are you sure you have `pip==20.1`? I'm unable to reproduce the error for this version. I do, however, see this for `pip>=21`. It seems that `poetry add light-the-torch` does **not** resolve the dependencies properly, because `light-the-torch` specifies

https://github.com/pmeier/light-the-torch/blob/eff3ea97b3892d2f580699444c52f5594df81ff4/setup.cfg#L32-L33

**Comment by dmatos2012** [comment]
No, I am sorry, you are completely right. I did `pip --version`, and forgot that its just the pip version for my system, which is 20.1

However, looking at the actual pip used by poetry, its indeed what you expect, 
`pip --version #21.1.2`  

> With these 4 steps I now have a clear way of debugging this without ever needing to step into poetry land. I'll get right on it.

Awesome, thanks for teaching me tho. I thought it was a way more in-depth, including logs and everything. I appreciate the time!

**Comment by pmeier** [comment]
> Awesome, thanks for teaching me tho. I thought it was a way more in-depth, including logs and everything. I appreciate the time!

Glad I could help. Remember this later in your OSS journey if you encounter someone that also does not know this and help them out :slightly_smiling_face: 

---

After some further debugging, I found that `poetry` completely ignores the `pip` dependency. To test this I've included an `requests<2.20` dependency and this is resolved properly:

```
$poetry add -vvv git+https://github.com/pmeier/light-the-torch.git#poetry
Resolving dependencies...
   1: fact: poetry-ltt is 0.1.0
   1: derived: poetry-ltt
   1: fact: poetry-ltt depends on light-the-torch (rev poetry)
   1: selecting poetry-ltt (0.1.0)
   1: derived: light-the-torch (0.4.0.dev3+g752cf17 git rev poetry)
   1: fact: light-the-torch (rev poetry) depends on requests (<2.20)
   1: selecting light-the-torch (0.4.0.dev3+g752cf17 752cf17)
   1: derived: requests (<2.20)
PyPI: No release information found for requests-0.0.1, skipping
PyPI: No release information found for requests-0.12.01, skipping
PyPI: No release information found for requests-2.15.0, skipping
PyPI: 131 packages found for requests <2.20
   1: fact: requests (2.19.1) depends on chardet (>=3.0.2,<3.1.0)
   1: fact: requests (2.19.1) depends on idna (>=2.5,<2.8)
   1: fact: requests (2.19.1) depends on urllib3 (>=1.21.1,<1.24)
   1: fact: requests (2.19.1) depends on certifi (>=2017.4.17)
   1: selecting requests (2.19.1)
   1: derived: certifi (>=2017.4.17)
   1: derived: urllib3 (>=1.21.1,<1.24)
   1: derived: idna (>=2.5,<2.8)
   1: derived: chardet (>=3.0.2,<3.1.0)
PyPI: No release information found for certifi-0, skipping
PyPI: 21 packages found for certifi >=2017.4.17
PyPI: No release information found for urllib3-0.3, skipping
PyPI: No release information found for urllib3-0.3.1, skipping
PyPI: No release information found for urllib3-0.4.0, skipping
PyPI: No release information found for urllib3-0.4.1, skipping
PyPI: 3 packages found for urllib3 >=1.21.1,<1.24
PyPI: No release information found for idna-0.1, skipping
PyPI: 3 packages found for idna >=2.5,<2.8
PyPI: 3 packages found for chardet >=3.0.2,<3.1.0
   1: selecting urllib3 (1.23)
   1: selecting idna (2.7)
   1: selecting chardet (3.0.4)
   1: selecting certifi (2021.5.30)
   1: Version solving took 4.563 seconds.
   1: Tried 1 solutions.
```

At this point, I can't offer any further support, because this seems to be either a bug in `poetry` or intended behavior. In any case, I think it is better to raise this issue to them. If you do, you can link to this thread, so the maintainers don't have to do the debugging again. 

I'll close this for now. If there is new information and it turns out that I need to fix something feel free to reopen.
